src/input.rs

```diff
@@ -0,0 +1,191 @@
+use std::cmp::Ordering;
+
+use wgpu::winit;
+
+#[derive(Debug, Clone, Copy, PartialEq, Default)]
+pub struct InputState {
+    pub camera_pan_ground: [f32; 2],
+    pub camera_pan_screen: [f32; 2],
+    pub camera_rotate: [f32; 2],
+    pub camera_zoom: f32,
+    pub camera_zoom_steps: i32,
+    pub camera_reset_viewport: bool,
+    pub close_requested: bool,
+    pub import_requested: bool,
+    pub window_resized: Option<winit::dpi::LogicalSize>,
+}
+
+#[derive(Debug, Default)]
+pub struct InputManager {
+    lmb_down: bool,
+    rmb_down: bool,
+    shift_down: bool,
+    input_state: InputState,
+}
+
+impl InputManager {
+    pub fn new() -> Self {
+        Self {
+            lmb_down: false,
+            rmb_down: false,
+            shift_down: false,
+            input_state: InputState::default(),
+        }
+    }
+
+    pub fn input_state(&self) -> &InputState {
+        &self.input_state
+    }
+
+    pub fn start_frame(&mut self) {
+        self.input_state = InputState::default();
+    }
+
+    pub fn process_event(&mut self, ev: winit::Event) {
+        const MODIFIERS_NONE: winit::ModifiersState = winit::ModifiersState {
+            logo: false,
+            shift: false,
+            ctrl: false,
+            alt: false,
+        };
+
+        // FIXME: these should come in as parameters
+        let gui_captured_keyboard: bool = false;
+        let gui_captured_mouse: bool = false;
+
+        match ev {
+            winit::Event::WindowEvent { event, .. } => match event {
+                winit::WindowEvent::CloseRequested => {
+                    self.input_state.close_requested = true;
+                }
+                winit::WindowEvent::KeyboardInput { input, .. } => {
+                    let winit::KeyboardInput {
+                        virtual_keycode,
+                        state,
+                        modifiers,
+                        ..
+                    } = input;
+
+                    // We respond to some events unconditionally, even if GUI has focus.
+                    match (virtual_keycode, state, modifiers) {
+                        // Cmd+Q for macOS
+                        #[cfg(target_os = "macos")]
+                        (
+                            Some(winit::VirtualKeyCode::Q),
+                            winit::ElementState::Pressed,
+                            winit::ModifiersState {
+                                logo: true,
+                                shift: false,
+                                ctrl: false,
+                                alt: false,
```

Comment:
This is slightly incorrect. Even if you add `alt` to the mix, it should close the application. `ctrl` and `shift` are used for system specific actions (logout and close all opened apps respectively). I think this should be rewritten as:

```rust
winit::ModifiersState {
    logo: true,
    shift: false,
    ctrl: false,
    ..
}
```

There's also other (primary) way to quit MacOS app and that's `cmd + w`. It means "Close current window" (or tab). Since we'll most likely have just 1 window it's gonna be identical to `cmd + q`. If we'd have multiple windows or tabs, we'll need to implement `cmd + shift + w` to close all windows / tabs as well and behaviour or `cmd + w` would change to close just active tab / window. `alt` behaviour of these shortcuts is the same as described above.

There's also an important difference between `cmd + w` and `cmd + q`. `cmd + w` keeps application loaded in memory. This can be observed by the app staying in dock and loading much faster. `cmd + q` kills the app. I don't know how to implement the former, but we can look into it (much) later. Either way a Trello task should be created with everything that's not gonna be currently implemented.

You can test all of this behaviour in some web browser or Finder (although it cannot be closed with `cmd + q`).

Reply:
I'll fix the `alt` here and create an issue for the rest.

Reply:
Done